Layering: formalize the "continent" concept

domenic opened the pull request:
This includes a couple fixes to help HTML and ES play well together. The second one, allowing the global this and global object to diverge, is significant. It fixes the following from HTML:

> This is a willful violation of the JavaScript specification current at the time of writing (ECMAScript edition 5, as defined in section 10.4.1.1 Initial Global Execution Context, step 3). The JavaScript specification requires that the this keyword in the global scope return the global object, but this is not compatible with the security design prevalent in implementations as specified herein.

/cc @annevk.

ljharb commented:
nice - if this goes in I'll be able to simplify the `System.global` proposal as well.

domenic commented:
I have added another commit defining "continents" (aka "vats") and simplifying initialization. It is very nice and simplifies some confusion I had with HTML integration immensely.

I could use some help figuring out how to make my `<dfn>`s into auto-links; I would like "continent" to link to my dfn wherever it appears.

anba commented:
I don't understand how the simplified initialization is going to work: The new Realm is not yet created, but the host environment should already provide the globalObject and globalThis value. And to be able to create objects, you need an associated Realm.

8.3.5 GetGlobalObject should probably return the globalThis value or is there a difference between `unresolvedVariable = 0` and `this.unresolvedVariable = 0` in browsers? (GetGlobalObject is only used for unresolved references in 6.2.3.2 PutValue.)

9.2.1.2 OrdinaryCallBindThis: The function this-binding should use the globalThis value, shouldn't it? (`(function(){ return this })() === this` in browsers.)

domenic commented:
> And to be able to create objects, you need an associated Realm.

Can you explain why that is? My understanding is that functions have an associated realm, but not objects. And IIUC this order of operations is not how it works in browsers, so we should probably change the spec to accomodate if that's the case.

> GetGlobalObject
>
> OrdinaryCallBindThis

I think these are correct as-is but will produce test cases later to confirm for you.

anba commented:
IIUC realm and global creation happens at the same time in browsers (JS_NewGlobalObject in SM or Context::New in V8). I think it's just the CreateIntrinsics step which should be moved out of CreateRealm into InitializeHostDefinedRealm.

domenic commented:
Hmm, so I didn't see you respond to my question of why you require a Realm to create objects. So I am going to assume that is OK?

> GetGlobalObject

This should continue to return the global object, not the global this value. The reasoning is that `var` declarations create non-configurable properties on the global object. But, in order to preserve the fundamental invariants, any declared vars must disappear from the window proxy object after navigation. ([Example](http://plnkr.co/edit/z8KIglzvM3bQIifP6kgb?p=preview).) Thus when observing the window proxy properties from the outside, they must appear non-configurable. And, it must not be possible to define a non-configurable property on the window proxy.

This is captured in the [tentative window proxy spec](https://github.com/domenic/window-proxy-spec). If GetGlobalObject used the thisValue, it would get a window proxy, and then defining any vars would fail, since you can't call [[DefineOwnProperty]] with Configurable: false. ([GlobalDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-globaldeclarationinstantiation) passes D = false for global var declarations.)

So GetGlobalObject should continue to use the actual global object, which it can define such properties on.

> OrdinaryCallBindThis

Correct, this should be changed. Uploading amended patch now.

anba commented:
> Hmm, so I didn't see you respond to my question of why you require a Realm to create objects. So I am going to assume that is OK?

As long as no intrinsics are needed, for example %ObjectPrototype% on the proto-chain. Alternatively the prototype chain needs to be updated after the intrinsics are present. (This seems to be the case [in SpiderMonkey](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/JSAPI_reference/JS_NewGlobalObject)).

>> GetGlobalObject
> 
> This should continue to return the global object, not the global this value.

GetGlobalObject is only used in PutValue, GlobalDeclarationInstantiation directly works on the global Environment Record. Nevertheless, I don't think the difference between `this.unresolvedVariable = 0` and `unresolvedVariable = 0` matters, it's more or less the same as the difference between `this.globalVariable = 0` and `globalVariable = 0` (the former calls Set() on the globalThis value whereas the latter calls Set() on the global object).

domenic commented:
> As long as no intrinsics are needed, for example %ObjectPrototype% on the proto-chain.

That makes sense. It would probably be easier to move those steps back into the middle of the algorithm as call-outs to the host. I'll do that then.

> Nevertheless, I don't think the difference between this.unresolvedVariable = 0 and unresolvedVariable = 0 matters

I think I end up agreeing. This is largely because WindowProxy transparently proxies [[Set]]. If it did not then we would have an observable difference. In such a hypothetical world I still think global object makes more sense than global this value.

anba commented:
Obviously I forgot to mention why the realm creation and initialization steps are currently split into several operations: These are all leftovers from previous drafts which contained the Realm API. That also means if the Realm API has a comeback, we will again need to split the initialization process into several methods. But I'm not sure it's worth to consider this aspect right now.

bterlson commented on spec.html:
```diff
@@ -6241,20 +6254,36 @@
+
+    <p>Each continent contains a set of job queues, as defined in <emu-xref href="#sec-jobs-and-job-queues"></emu-xref>.</p>
+
+    <emu-note>Typically, a continent is implemented as a single thread of execution. For example, in a web browser, each unit of related similar-origin browsing contexts consists of a single continent, as does each worker.</emu-note>
```
This note doesn't clarify anything to me. In general, requiring super detailed domain knowledge of other specifications to understand what this note is saying seems bad.

domenic replied:
Hmm. This doesn't seem like super-detailed knowledge. Do you think there's a way to phrase this to make it easier? Maybe replace "unit of related similar origin browsing contexts" with "browser tab (including any frames or iframes)"?

bterlson replied:
If someone is not a browser implementer, this example would require much research to understand I think? I'm not sure what you're trying to clarify here, though, so it's hard for me to suggest alternatives.

domenic replied:
I'm trying to provide a somewhat-intuitive conceptual understanding for authors and implementers as to what a continent corresponds to in the "real world." Before embarking on this journey, it wasn't clear to me at all whether we were supposed to spin up a separate instance of the ES spec for each global, or each browser, or each browser window, or what. I am trying to point out that it's essentially one continent (or "instance of the ES spec") per thread/event loop. The "For example" tries to bring that even closer to real-world experience, showing a system where many realms can coexist (with "political barriers" between them, cf. [es-discuss](https://esdiscuss.org/topic/holy-mixed-metaphors-batman-was-event-loops-in-navigated-away-from-windows#content-2)) within the same continent.

bterlson replied:
This seems like an appropriate clarification for HTML but I don't think it provides much value in the ECMAScript spec and has the downside of being potentially confusing.

domenic replied:
I think it'd be helpful if ES were able to give some example of what a continent was in the real world of actual JS engine implementations. Can you suggest any such examples that you don't think would be confusing?

bterlson commented on spec.html:
```diff
@@ -6241,20 +6254,36 @@
+
+    <p>The host environment must create a continent prior to the execution of any Jobs or the evaluation of any ECMAScript code.</p>
+
+    <p>A continent may contain multiple realms, as defined in <emu-xref href="#sec-code-realms"></emu-xref>. Each realm belongs to exactly one continent. Depending on the host environment, code executed within a given continent's realms may or may not be able to reach objects from other realms in that continent. However, code executed in one continent must not be able to directly reference or sha ...
```
This paragraph creates a normative requirement that continents can't share or reference with each-other. Doesn't this invalidate the opening sentence that continents are the largest unit the specification deals with (since it is now prescribing how the planet's continents may communicate)? Also the terms "share" and "reference", are they defined rigorously already? In general, it seems ok to omit this paragraph and leave inter-continent communication up to host discretion.

domenic replied:
I agree there's a bit of a contradiction here. I'd argue it's OK, maybe with a slight tweak to the first sentence. In general it seems OK for this sentence to "overreach" a bit, using undefined terms to convey an intention. The sentence is almost entirely drawn from @allenwb's https://esdiscuss.org/topic/event-loops-in-navigated-away-from-windows#content-17

But, happy to remove, especially if it'll help get things in faster.

bterlson replied:
Let's see if @allenwb or @erights have comments on it.

bterlson commented on spec.html:
```diff
@@ -6211,6 +6222,8 @@
         1. Return NormalCompletion(~empty~).
       </emu-alg>
+
+      <emu-note>For host environments consisting of multiple continents, when those host environments enqueue their own jobs, they should generally specify which continent's job queues they are enqueuing on.</emu-note>
```
This note confuses me. How would it be possible to enqueue a job, which involves executing code, which requires a realm, which is on a continent, without specifying a continent or specific realm within a continent?

domenic replied:
EnqueueJob takes neither _realm_ nor _continent_ parameters. This is trying to clarify that there's an implicit continent parameter (as there is with essentially everything in the spec, but most of the time it is not as relevant). In particular, I had first written the HTML spec calling out to EnqueueJob("ScriptJobs", ScriptEvaluationJob, « _sourceText_ ») but the results were very confusing since it wasn't clear that there was actually one "ScriptJobs" queue per continent, i.e. it wasn't clear what relative order needed to be preserved between.

Note that executing code does _not_ in fact require a realm.

domenic replied:
OK, I got this a bit wrong, I see. PendingJobs have a [[Realm]], and you could argue that the job queue chosen is implicitly the one belonging to the [[Realm]]'s parent continent. The [[Realm]] is chosen based on the running execution context. I could revise this note to something like

> For host environments consisting of multiple continents, each with their own set of job queues, it can be helpful to realize that EnqueueJob always enqueues in a job queue belonging to the running execution context's realm's continent.

or even just

> EnqueueJob always enqueues in a job queue belonging to the running execution context's realm's continent.

The main thing I am clarifying here (indeed, the main thing I am clarifying with this whole PR) is that in a host environment with multiple continents, I want it to be obvious where EnqueueJob("ScriptJobs", ...) is actually happening.

Now I need to figure out how to set the running execution context from within HTML... should be doable.

bterlson replied:
> I want it to be obvious where EnqueueJob("ScriptJobs", ...) is actually happening

Maybe it's obvious already? EnqueueJob calling some other Continent doesn't seem a likely implementation alternative...

domenic replied:
Maybe. I can give it a try. It's certainly not obvious when reading the spec that there's an implicit "current execution context" -> implicitly created PendingJob's [[Realm]] field -> continent link.

bterlson replied:
Can you expand on that last bit? My mental model is more like, there is a running execution context which has a realm and continent and that running execution context must be the same prior to and after the EnqueueJob call. Am I missing the problem?

domenic replied:
Without continents:

- HTML has event loops, each of which has an ECMAScript implementation, each of which has a set of job queues.
- When you want to run some script, you find an appropriate execution context within that event loop, set it to the running execution context. Then you call EnqueueJob("ScriptJobs", ScriptEvaluationJob, « sourceText »).
- From the running execution context, the ES spec will create a PendingJob record whose [[Realm]] is set appropriately. From that [[Realm]], we infer (although it is never explicitly stated) that the appropriate "ScriptJobs" queue to enqueue ScriptEvaluationJob into is that of the ECMAScript implementation containing pendingJobRecord.[[Realm]].

With continents:

- HTML has event loops, which are associated one-to-one with ES continents.
- When you want to run some script, you find an appropriate execution context within that event loop, set it to the running execution context. Then you call EnqueueJob("ScriptJobs", ScriptEvaluationJob, « sourceText »), noting that the "ScriptJobs" queue will be the one belonging to the event loop's continent.

Maybe it is not a big deal, and people can figure out the first version. The second version makes it a lot clearer to me though.

bterlson replied:
I am confused at how you could arrive at any other interpretation still so I don't see the note as necessary. Even externally, it seems obvious that the continent whose execution context you activate in order to call EnqueueJob will be the continent whose job queues you are appending to.

domenic replied:
It's not explicitly stated anywhere, that's for sure. It's never stated which continent's job queues are appended to at all, because the spec doesn't currently have a notion of continent, and only deals with a single set of job queues.

bterlson replied:
I was reviewing this in the context of the following continent text being present... if that's not present, the note should be removed on the grounds that "continent" is not defined.

domenic replied:
I don't think simply defining continent serves to make it clear which job queues NextJob enqueues in.

bterlson commented on spec.html:
```diff
@@ -5219,6 +5219,17 @@
+                Object
+              </td>
+              <td>
+                The value returned by `this` in global scope. This may be the same as the global object, or may not. For example, in HTML the `this` value is a `WindowProxy` object, whereas the global object is a `Window`.
```
I don't think this example is required or illustrative for ECMAScript spec readers. I would say something like:

> The value returned by `this` in global scope. This is usually the same object as the global object, however hosts may provide any ECMAScript Object value.

Has the added benefit of making it a normative requirement that the this value be an object, which I think is good :)

domenic replied:
This was kind of in the spirit of "for example, in the HTML document object model the window property of the global object is the global object itself." Happy to drop it, but it seems kind of nice. Also happy to do require it to be an Object, although that is implicitly required by InitializeHostDefinedRealm's latest definition. (In general these descriptions seem non-normative to me...)

domenic replied:
Also, I'm not sure "usually the same as" is a good sentence fragment given that in reality 95% of InitializeHostDefinedRealm() calls will use divergent global and `this` value. (I think an estimate of 5% for Node.js is generous personally...)

bterlson replied:
On reflection I agree that the "usually" clause is bad. Just remove it. It adds as much information as the "It may... or may not" clause :-P

bterlson commented on spec.html:
```diff
@@ -5427,7 +5438,7 @@
         <emu-clause id="sec-global-environment-records-hasthisbinding">
           <h1>HasThisBinding ()</h1>
-          <p>Global Environment Records always provide a `this` binding whose value is the associated global object.</p>
+          <p>Global Environment Records always provide a `this` binding.</p>
```
I don't think this copy is required anymore.

bterlson commented on spec.html:
```diff
@@ -5893,15 +5903,16 @@
-      <h1>SetRealmGlobalObject ( _realmRec_, _globalObj_ )</h1>
-      <p>The abstract operation SetRealmGlobalObject with arguments _realmRec_ and _globalObj_ performs the following steps:</p>
+      <h1>SetRealmGlobalObject ( _realmRec_, _globalObj_, _thisValue_ )</h1>
+      <p>The abstract operation SetRealmGlobalObject with arguments _realmRec_, _globalObj_, and _thisValue_ performs the following steps:</p>
```
If thisValue is optional, follow the associated notational conventions (brackets around the signature). Also I think the signature is enough, the following paragraph adds nothing (it's useful for some of the old-style abstract ops that don't use the signature notation).

domenic replied:
It's not actually optional; all places in the spec call it with all three arguments. Sometimes *undefined* is used as a default-value signifier.

domenic replied:
This is because it's quite hard to get ES-style compositional default values in ECMASpeak. If we stuck with "not passed," the result would be that InitializeHostDefinedRealm() would have to branch and call out to either SetRealmGlobalObject(realmRec, globalObj) or SetRealmGlobalObject(realmRec, globalObj, thisValue). In other words, **undefined** and not passed are not currently equivalent in ECMASpeak.

bterlson replied:
Yes, I see. For some reason I think I read "not passed" when I made the original comment :(

bterlson commented:
@Domenic I left some specific comments, but in general I want to split this into two pieces. The globalThis thing is clearly a layering concern that I am happy to pull in now. The text around "continents" I worry is subtle enough and normative enough that it should probably advance through the normal proposal process. It should see some committee debate :)

domenic commented:
> The globalThis thing is clearly a layering concern that I am happy to pull in now.

Awesome!

> The text around "continents" I worry is subtle enough and normative enough that it should probably advance through the normal proposal process. It should see some committee debate :)

Hmm. This wasn't my intent, although I do think it would be a good thing to highlight to the committee and let them help advise on changes to. Could we make it less normative perhaps? Perhaps by removing the object references sentence you noted? Would it help to get sign-off from committee members with particularly relevant interests, like @allenwb and @erights?

My main reason for introducing it was that I need something that HTML can stick on each HTML event loop to make it clear that each HTML event loop gets a single set of ES job queues. Currently the only term ES provides for that is "ECMAScript implementation," as in, "Every ECMAScript implementation has at least the Job Queues defined in Table 26." But using this term seems misleading in a number of ways, especially because the spec _usually_ uses it in the sense of "JS engine". (Example: "The "callee" property has a more specific meaning for non-strict functions and a "caller" property has historically been provided as an implementation-defined extension by some ECMAScript implementations.")

That  is, the main reason for introducing it is that I need some vague concept that HTML can:
- Create whenever the browser creates a new event loop
- Reference whenever it does EnqueueJob(...), in order to make it clear what collection of realms ("continent") the given script execution or promise microtask is executing in.

"Continent" seemed like a nice way to get this while also making at least a brief start toward the stuff outlined in the linked es-discuss threads.

bterlson commented:
I understand the intent and I completely agree with the rationale. I also agree that the normal proposal process is overweight. It just seems to me like a subtle enough area that we should discuss more, and I think is something we would typically discuss in TC39 before including in a draft. That said, we definitely want feedback from @allenwb and @erights before pulling this in so let's get that first.

domenic commented:
Sure, I guess I can try to work around the lack in HTML for the next couple months. I kind of think since it's non normative it should be fine to commit first ask questions later but I understand.

domenic commented:
I addressed the review comments on the second commit, so those first two should be pull-in-able, I think. (I also rebased on master.) I left the third one in this branch because I didn't want to lose the conversations.

domenic commented:
Actually, wait, I need to keep the merger of Initialization() and InitializeHostDefinedRealm(), even without the continent stuff. Let me separate out the continent commit into two, one that you can pull in now and one that we can leave for later.

domenic commented:
OK, all separated. This should make it easier for @allenwb and @erights (and anyone else) to review just the last commit about continents, if they want.

zenparsing commented:
I would tend to agree that any specification of vats within 262 needs to have some tc39-time.

bterlson commented:
I merged the uncontroversial aspects. Continent work is now pending consensus.

domenic commented:
I rebased this on master, and renamed the PR, so that it only contains the "continent" concept now. The previous commits were merged in b1ef33bf40a324dcd8cd92d2bcc6e08c7074978d, 9474c58c6baf54e8b5347a182c2fa6ee10dab071, and eacbc9d06cfc453a51d22d7980ec893a5c7e2de0.

This turns out to not be as necessary for HTML's clarity as I thought, given that HTML now overrides EnqueueJob and NextJob, so I don't feel too strongly about this addition any more. But, I think it's kind of a nice start at speccing something TC39 would probably like to spec anyway. It makes it clearer to readers, IMO, how a host environment with multiple "threads" would work, and fit with the existing specification. So I'll leave it up for brief discussion next meeting.

In the meantime, I'd invite @allenwb to take a look and give his thoughts, since the original idea is largely due to him.

jmdyck commented on spec.html:
```diff
@@ -6285,6 +6287,20 @@
+
+    <p>A continent may contain multiple realms, as defined in <emu-xref href="#sec-code-realms"></emu-xref>. Each realm belongs to exactly one continent. Depending on the host environment, code executed within a given continent's realms may or may not be able to reach objects from other realms in that continent. However, code executed in one continent must not be able to directly reference or sha ...
+
+    <p>Each continent contains a set of job queues, as defined in <emu-xref href="#sec-jobs-and-job-queues"></emu-xref>.</p>
```
Presumably each continent also contains a single execution context stack.

bterlson commented:
Going to close this PR for now as the continent formalization has moved to the shared memory proposal.

lars-t-hansen commented:
For reference, work continues in https://github.com/tc39/ecmascript_sharedmem/issues/27.
